### ISSUE TITLE
$env:CMAKE_ARGS='-DLLAMA_CUBLAS=on' issue

### ISSUE THREAD

**Issue by Zetto12** [imartinez/privateGPT#1291]
I hope someone can help me. Got privateGPT working but I'm stuck on enabling GPU support. I particular at this step:

$env:CMAKE_ARGS='-DLLAMA_CUBLAS=on'; poetry run pip install --force-reinstall --no-cache-dir llama-cpp-python

I get the error 'The filename, directory name, or volume label syntax is incorrect.'

GPU is Nvidia 2070

![image](https://github.com/imartinez/privateGPT/assets/95339767/0c8383fc-aaeb-42fb-b98b-d65a7b3c9fb6)

Thanks,
Martin

**Comment by Kaotic3** [comment]
Hey Martin,

You may want to check your environment variables and ensure that you have these activated there, I had some problems setting it via the cmd.

I have as variables in Environment Variables settings:

CMAKE_ARGS  -DLLAMA_CULBAS=on

CUDA_PATH
CUDA_PATH_V12_2
FORCE_CMAKE = 1
LLAMA_CLBLAST = 1

Also in Path for System I have the path for the NVIDIA GPU computing toolkit.

This seems to work at least for my GPU.

**Comment by Kaotic3** [comment]
Once you set them as variables - you don't need to include them in the command to run privateGPT.

The error message you are reporting sounds a lot like the error message I get when I try to use 'git clone' before I have installed git - where git is not recognised.

In the command line I use:

`poetry run python -m private_gpt`

I don't include anything else, because I have already set it.  It runs fine and utilises my GPU.

**Comment by EmmaWebGH** [comment]
Same problelm
Tried both
`$env:CMAKE_ARGS='-DLLAMA_CUBLAS=on'; poetry run pip install --force-reinstall --no-cache-dir llama-cpp-python`
and
`CMAKE_ARGS='-DLLAMA_CUBLAS=on'; poetry run pip install --force-reinstall --no-cache-dir llama-cpp-python`
because I thought maybe $env was like "the name of your env" or "make sure you've activated your venv" or something
Anyway, both give 
```
'CMAKE_ARGS' is not recognized as an internal or external command,
operable program or batch file.
```

CMAKE is installed because in my anaconda shell I can type cmake and it returns usage, or I can run cmake --help for example.
So then why does the command that the docs tells me to copy/paste, not work please?

**Comment by Kaotic3** [comment]
If you have your variables set (you can check in windows by typing in Environment Variables in the search bar, from memory Echo in Linux but ask GPT) - then you don't need the $env: commands anymore.  You just need the install commands.  poetry run pip install....

The install commands will run from anaconda.

Usually when there is such a "wheel" error, you have to scroll all the way back up to the top to see what the original error was that caused the issue.

If you can get back into conda by setting your variables, then you can try uninstalling llama-cpp-python and then install a specific version of it.

I believe from memory, but go google to be sure, that if you are using GGUF model files it is =>0.1.8.3 this is what I use  but look into it to see which one you need.

**Comment by EmmaWebGH** [comment]
Thanks @Kaotic3 but the command in question is to enable CUDA/GPU, as standard install with GGUF/CPU is very slow.
So if those commands are not needed any more, how to activate GPU and get BLAS=1?

**Comment by Kaotic3** [comment]
Yes, but the command is creating an Environment Variable that is what $env is doing.

If you create them as environment variables, you no longer need to invoke them every time and they are set for future use.  Then you can just move to working out the issue you are having with llama-cpp-python / nvidia.

I am assuming that you have installed the Nvidia Toolkit? You have 12.2 or 12.3?

I noticed the last time that I did this that order made a difference.  I had already, I think from memory, installed the correct llama-cpp but then installed Nvidia's toolkit afterward - and could not get BLAS=1 working, until I uninstalled and reinstalled llama.

Yes, Envs are annoying, but it is a good way to keep all these messy installs separate.  There will be times where you will need to keep an older version of something running to keep something working, but you don't want that to dictate your entire python experience and everything else you do or can do with python.

**Comment by EmmaWebGH** [comment]
Thanks, I understand the purpose of envs I was just commenting on the number of Llama installs I am accruing.

I installed Nvidia first, then uninstalled / reinstalled lama-cpp-python.

I got the llama-cpp--python wheel to build finally, by doing this:
In Anaconda Prompt:
```
pip uninstall -y llama-cpp-python
set CMAKE_ARGS="-DLLAMA_CUBLAS=on"
set FORCE_CMAKE=1
pip install llama-cpp-python==0.1.57 --no-cache-dir
```
But my BLAS=0 still

There is no message or echo of anything when I set the args. Should there be?
Now that wheel is built, how do I get this running on GPU?

To be honest I don't understand why it's not set to install with GPU from the get go, and if someone doesn't have a GPU it can fall back to CPU...

**Comment by Kaotic3** [comment]
if after this:

`set CMAKE_ARGS="-DLLAMA_CUBLAS=on"`

You do `echo %CMAKE_ARGS%` - and you get nothing. Then you have not set the variables.

If you are on windows simple type in "Environment Variables" in the search bar, and open up the environment variable control and follow the post I made to Martin previously to create all the variables you need.

I created variables for both User and System.

That is also quite an old version of llama - is this because you are trying to use particular Model types and have been told to use that particular llama? I was using the most up to date version on Thursday last week and it worked fine, so not sure that a particular version is necessary unless you are working with particular models.

**Comment by EmmaWebGH** [comment]
Oh, I totally didn't catch the old version, thanks for pointing that out. I was following solutions from other users posted on a gist, not sure why they were using that version.
I will try again today with current version and your suggestions and report back, Thanks!

**Comment by EmmaWebGH** [comment]
Installed current llama-cpp. Set args via conda command prompt.
When I echo:
 `echo %CMAKE_ARGS%`
I get `"-DLLAMA_CUBLAS=on"`. 

Seems like success?
BUT when I `make run`, still BLAS=0
`AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | SSSE3 = 0 | VSX = 0 |
`

Tried closing anaconda prompt and reopening, echoed args, still says it's on, but again make run starts up with BLAS=0.

**Comment by EmmaWebGH** [comment]
I've set environment variables for windows and it's still BLAS=0 on startup.
Screenshot. This is set for both user and system. Have I done this wrong?

![SystemPropertiesAdvanced_olPxcaFmNf](https://github.com/imartinez/privateGPT/assets/123797054/834fbcd1-b1b9-4df6-984b-f1141ff3a601)

**Comment by Kaotic3** [comment]
One question - are those variables set for both the User and System?  Mine are set for both, whether that is necessary, I don't actually know...

The last steps that I can think of are:

1. Ensure you have the Cuda tools in your PATH variable and also have: CUDA_PATH and CUDA_PATH_V12_2 in your environment variables, where the V12_2 is the version that you are using.  This appears in both my user and system variables (though PATH is for system only).

2. Uninstall llama-cpp and reinstall llama-cpp after verifying the above, it may be that you need to have the variables set before installing.

Beyond that, I don't know how else to assist.

**Comment by EmmaWebGH** [comment]
Yes, as above I wrote both user and system.

Oh it needs CUDA 12.2 does it?
OK that might be the problem. I have 12.0. Alright then...

**Comment by meuerdesign** [comment]
> this is solved: [abetlen/llama-cpp-python#721 (comment)](https://github.com/abetlen/llama-cpp-python/issues/721#issuecomment-1722280380)

This is the right answer.
I simply had to reinstall CUDA and it worked.
